### DISCUSSION TITLE
evbuffer_read syscall overhead

### DISCUSSION THREAD

**Discussion by yydcool** [transmission/transmission#3504] [General]
I did a profiling with 4.0 beta version that shows ~40% of transmission's CPU usage spent on syscall when downloading at ~30MB/s with TCP only (no UTP).
It turns out most syscall overhead is from
peer-io.cc:
event_read_cb()
  --> evbuffer_read()
As I read libevent 2.1.x code, evbuffer_read() only reads at most 4KB at a time and issues at least 2 syscall: a ioctl() and a readv()
I wonder if there's a good way to improve this.
libevent 2.2(unstable) has added the logic to increase that 4KB limit, but as I tested it won't help much. The reason is that the event "socket have some bytes arrived to read" issued too soon, by the time event_read_cb() is called, only a few KB is available (while we are expecting a whole piece of X MB to come).

**Comment by ckerr** [comment]
@yydcool this is very interesting stuff; thank you for looking into this!

I have been considering switching to asio; however, it will probably not happen soon as use of libevent is so extensive in Transmission (e.g. it's used for the RPC server, for DNS, and evbuffer is used extensively). So if you have any ideas on how things could be improved in the interim I'd love any suggestions or patches that you might have; e.g. code to upstream into libevent or even a custom `tr_evbuffer_read()` that could be in libtransmission.

I'd love to hear more about making peer-io's socket io faster; but as a side question, is this with MSE / encryption disabled? I'd expect arc4 to be using up a decent amount of CPU as well.

**Comment by ckerr** [comment]
Breadcrumb: I believe https://github.com/libevent/libevent/commit/8c2001e92a6615cc7b4efc8e33cff58c28effd30 is what yydcool is referring to in libevent 2.2

**Reply by yydcool** [reply]
Yes, that is the patch to add tunable max read buffer size.
I do have encryption disabled, then I guess after encryption enabled this overhead really be a minor issue .
I've run another test with enforcing my x86 CPU to run at 800Mhz to simulate a low performance device. Without encryption, and without tuning max read size(so it is still 4KB), transmission uses ~70% CPU@800Mhz to download at ~30MB/s. So I guess it is still fine ;-)

**Comment by ckerr** [comment]
I'm very interested in improving performance at high transfer speeds, but in practice I don't have much opportunity to do that kind of tuning. If you want to continue work on this and have any patches, I'd be happy to review them.
